File: src/omega/omega_apis.cc
#include <iostream>
#include <chrono>

#include <dgl/runtime/container.h>
#include <dgl/runtime/object.h>
#include <dgl/packed_func_ext.h>
#include "../c_api_common.h"

#include "distributed_block.h"
#include "distributed_sampling.h"
#include "partition_request.h"
#include "sampling.h"
#include "trace_gen_helper.h"
#include "trace.h"
#include "facebook_dataset_partition.h"

using namespace dgl::runtime;

namespace dgl {

namespace omega {

std::pair<HeteroGraphPtr, IdArray> ToBlockGPU(const IdArray& u,
                                              const IdArray& v,
                                              const IdArray& dst_ids,
                                              const IdArray& src_ids,
                                              const IdArray& new_lhs_ids_prefix);

};

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaToDistributedBlocks")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    int num_machines = args[0];
    int machine_rank = args[1];
    int num_gpus_per_machine = args[2];
    IdArray target_gnids = args[3];
    IdArray src_gnids = args[4];
    IdArray src_part_ids = args[5];
    IdArray dst_gnids = args[6];

    auto results = omega::ToDistributedBlocks(
        num_machines,
        machine_rank,
        num_gpus_per_machine,
        target_gnids,
        src_gnids,
        src_part_ids,
        dst_gnids);

    CHECK_EQ(num_gpus_per_machine, results.first.size());
    CHECK_EQ(num_gpus_per_machine, results.second.size());

    List<runtime::Value> ret_list;
    for (int i = 0; i < num_gpus_per_machine; i++) {
      ret_list.push_back(runtime::Value(MakeValue(HeteroGraphRef(results.first[i]))));
      ret_list.push_back(runtime::Value(MakeValue(results.second[i])));
    }

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaToDistributedBlock")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    int num_machines = args[0];
    int machine_rank = args[1];
    int num_gpus_per_machine = args[2];
    int local_gpu_idx = args[3];
    IdArray target_gnids = args[4];
    IdArray src_gnids = args[5];
    IdArray src_part_ids = args[6];
    IdArray dst_gnids = args[7];

    auto result = omega::ToDistributedBlock(
        num_machines,
        machine_rank,
        num_gpus_per_machine,
        local_gpu_idx,
        target_gnids,
        src_gnids,
        src_part_ids,
        dst_gnids);

    List<runtime::Value> ret_list;
    ret_list.push_back(runtime::Value(MakeValue(HeteroGraphRef(result.first))));
    ret_list.push_back(runtime::Value(MakeValue(result.second)));

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaToBlock")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    // auto start_time = std::chrono::high_resolution_clock::now();

    const HeteroGraphRef empty_graph_ref = args[0];
    IdArray u = args[1];
    IdArray v = args[2];
    IdArray dst_ids = args[3];
    IdArray src_ids = args[4];
    IdArray new_lhs_ids_prefix = args[5];

    runtime::List<runtime::ObjectRef> ret_list;

    auto device_type = u->ctx.device_type;
    CHECK_EQ(v->ctx.device_type, device_type);
    CHECK_EQ(dst_ids->ctx.device_type, device_type);
    CHECK_EQ(new_lhs_ids_prefix->ctx.device_type, device_type);

    if (src_ids->shape[0] > 0) {
      CHECK_EQ(src_ids->ctx.device_type, device_type);
    }

    if (device_type == DGLDeviceType::kDGLCPU) {
      auto ret = omega::ToBlock(empty_graph_ref, u, v, dst_ids, src_ids, new_lhs_ids_prefix);
      ret_list.push_back(HeteroGraphRef(ret.first));
      ret_list.push_back(runtime::Value(MakeValue(ret.second)));
      *rv = ret_list;
    } else {
      CHECK_EQ(device_type, DGLDeviceType::kDGLCUDA);
      auto ret = omega::ToBlockGPU(u, v, dst_ids, src_ids, new_lhs_ids_prefix);
      ret_list.push_back(HeteroGraphRef(ret.first));
      ret_list.push_back(runtime::Value(MakeValue(ret.second)));
      *rv = ret_list;
    }

    // std::chrono::duration<double> elapsed_time = std::chrono::high_resolution_clock::now() - start_time;
    // std::cerr << "_CAPI_DGLOmegaToBlock " << elapsed_time.count() << " seconds" << std::endl;

  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaPartitionRequest")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    int num_machines = args[0];
    int num_gpus_per_machine = args[1];
    IdArray nid_partitions = args[2];
    IdArray num_assigned_targets_per_gpu = args[3];
    IdArray target_gnids = args[4];
    IdArray src_gnids = args[5];
    IdArray dst_gnids = args[6];

    auto ret = omega::PartitionRequest(
      num_machines,
      num_gpus_per_machine,
      nid_partitions,
      num_assigned_targets_per_gpu,
      target_gnids,
      src_gnids,
      dst_gnids);

    runtime::List<runtime::Value> ret_list;
    for (const auto& s : ret.first) {
      ret_list.push_back(runtime::Value(MakeValue(s)));
    }

    for (const auto& d : ret.second) {
      ret_list.push_back(runtime::Value(MakeValue(d)));
    }

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaTraceGenHelper")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    int first_new_gnid = args[0];
    IdArray infer_target_mask = args[1];
    IdArray batch_local_ids = args[2];
    IdArray u = args[3];
    IdArray v = args[4];
    IdArray u_in_partitions = args[5];
    IdArray v_in_partitions = args[6];
    bool independent = args[7];

    auto ret = omega::TraceGenHelper(
      first_new_gnid,
      infer_target_mask,
      batch_local_ids,
      u,
      v,
      u_in_partitions,
      v_in_partitions,
      independent);

    runtime::List<runtime::Value> ret_list;
    for (int i = 0; i < 3; i++) {
      ret_list.push_back(runtime::Value(MakeValue(ret[i])));
    }

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.dist_sample._CAPI_DGLOmegaSplitLocalEdges")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    int num_machines = args[0];
    int machine_rank = args[1];
    int num_gpus_per_machine = args[2];
    int gpu_rank = args[3];
    IdArray global_src = args[4];
    IdArray global_dst = args[5];
    IdArray global_src_part_ids = args[6];

    auto ret = omega::SplitLocalEdges(
      num_machines,
      machine_rank,
      num_gpus_per_machine,
      gpu_rank, 
      global_src, 
      global_dst, 
      global_src_part_ids);

    int num_gpus = num_machines * num_gpus_per_machine;
    runtime::List<runtime::Value> ret_list;
    for (int i = 0; i < num_gpus; i++) {
      ret_list.push_back(runtime::Value(MakeValue(ret.first[i])));
    }

    for (int i = 0; i < num_gpus; i++) {
      ret_list.push_back(runtime::Value(MakeValue(ret.second[i])));
    }

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaSampleEdges")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    IdArray target_gnids = args[0];
    IdArray src_gnids = args[1];
    IdArray dst_gnids = args[2];
    IdArray fanouts_array = args[3];
    const auto& fanouts = fanouts_array.ToVector<int64_t>();

    auto ret = omega::SampleEdges(
      target_gnids,
      src_gnids,
      dst_gnids,
      fanouts);

    runtime::List<runtime::Value> ret_list;
    for (auto p : ret) {
      ret_list.push_back(runtime::Value(MakeValue(p.first)));
      ret_list.push_back(runtime::Value(MakeValue(p.second)));
    }

    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaGetCppTraces")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    runtime::List<runtime::Value> ret_list;
    for (auto t : omega::GetCppTraces()) {
      ret_list.push_back(runtime::Value(MakeValue(std::get<0>(t))));
      ret_list.push_back(runtime::Value(MakeValue(std::get<1>(t))));
      ret_list.push_back(runtime::Value(MakeValue(std::get<2>(t))));
    }
    *rv = ret_list;
  });

DGL_REGISTER_GLOBAL("omega.omega_apis._CAPI_DGLOmegaPartitionFacebookDataset")
  .set_body([](DGLArgs args, DGLRetValue* rv) {
    const int num_parts = args[0];
    const std::string input_dir = args[1];
    const List<Value> edge_file_paths_values = args[2];
    const bool include_out_edges = args[3];
    const double infer_prob = args[4];
    const int num_omp_threads = args[5];

    std::vector<std::string> edge_file_paths;
    for (int i = 0; i < edge_file_paths_values.size(); i++) {
      edge_file_paths.push_back(edge_file_paths_values[i]->data);
    }

    omega::PartitionFacebookDataset(num_parts, input_dir, edge_file_paths, include_out_edges, infer_prob, num_omp_threads);
  });

}

File: python/dgl/omega/omega_apis.py
import torch

import dgl

from .distributed_block import DGLDistributedBlock
from ..heterograph import DGLBlock
from .._ffi.function import _init_api
from .. import backend as F
from .. import utils

empty_graph = dgl.graph(([], []))
empty_tensor = torch.tensor([])

def get_num_assigned_targets_per_gpu(num_machines, num_gpus_per_machine, num_targets):
    num_total_gpus = num_machines * num_gpus_per_machine
    num_targets_per_gpu = [0] * num_total_gpus
    for machine_idx in range(num_machines):
        num_targets_in_machine = num_targets // num_machines
        if machine_idx < num_targets % num_machines:
            num_targets_in_machine += 1

        for gpu_idx in range(num_gpus_per_machine):
            global_gpu_idx = machine_idx * num_gpus_per_machine + gpu_idx
            num_targets_per_gpu[global_gpu_idx] = num_targets_in_machine // num_gpus_per_machine
            if gpu_idx < num_targets_in_machine % num_gpus_per_machine:
                num_targets_per_gpu[global_gpu_idx] += 1
    return num_targets_per_gpu

def to_distributed_blocks(
    num_machines,
    machine_rank,
    num_gpus_per_machine,
    target_gnids,
    src_gnids,
    src_part_ids,
    dst_gnids):
    ret = _CAPI_DGLOmegaToDistributedBlocks(
        num_machines,
        machine_rank,
        num_gpus_per_machine,
        F.zerocopy_to_dgl_ndarray(target_gnids),
        F.zerocopy_to_dgl_ndarray(src_gnids),
        F.zerocopy_to_dgl_ndarray(src_part_ids),
        F.zerocopy_to_dgl_ndarray(dst_gnids))

    num_assigned_target_nodes = get_num_assigned_targets_per_gpu(
        num_machines, num_gpus_per_machine, target_gnids.shape[0])
    dist_blocks = []
    gpu_ranks = [i for i in range(num_machines * num_gpus_per_machine)]
    for gpu_idx in range(num_gpus_per_machine):
        g_idx = ret[2 * gpu_idx]
        src_gnids_in_block = F.from_dgl_nd(ret[2 * gpu_idx + 1])
        gpu_rank_in_group = num_gpus_per_machine * machine_rank + gpu_idx

        dist_block = DGLDistributedBlock(
            gpu_rank_in_group,
            gpu_ranks,
            num_assigned_target_nodes,
            gidx=g_idx,
            ntypes=(['_N'], ['_N']),
            etypes=['_E'])

        assert dist_block.is_unibipartite
        dist_block.srcdata[dgl.NID] = src_gnids_in_block
        dist_blocks.append(dist_block)
    return dist_blocks

def to_distributed_block(
    num_machines,
    machine_rank,
    num_gpus_per_machine,
    local_gpu_idx,
    target_gnids,
    src_gnids,
    src_part_ids,
    dst_gnids,
    gpu_ranks=None):
    ret = _CAPI_DGLOmegaToDistributedBlock(
        num_machines,
        machine_rank,
        num_gpus_per_machine,
        local_gpu_idx,
        F.zerocopy_to_dgl_ndarray(target_gnids),
        F.zerocopy_to_dgl_ndarray(src_gnids),
        F.zerocopy_to_dgl_ndarray(src_part_ids),
        F.zerocopy_to_dgl_ndarray(dst_gnids))
    
    num_assigned_target_nodes = get_num_assigned_targets_per_gpu(
        num_machines, num_gpus_per_machine, target_gnids.shape[0])

    g_idx = ret[0]
    src_gnids_in_block = F.from_dgl_nd(ret[1])
    gpu_rank_in_group = num_gpus_per_machine * machine_rank + local_gpu_idx

    if gpu_ranks is None:
        gpu_ranks = [i for i in range(num_machines * num_gpus_per_machine)]

    dist_block = DGLDistributedBlock(
        gpu_rank_in_group,
        gpu_ranks,
        num_assigned_target_nodes,
        gidx=g_idx,
        ntypes=(['_N'], ['_N']),
        etypes=['_E'])

    assert dist_block.is_unibipartite
    dist_block.srcdata[dgl.NID] = src_gnids_in_block

    return dist_block

def to_block(u, v, dst_ids, src_ids=None):
    if src_ids == None:
        src_ids = empty_tensor

    graph_idx, src_orig_ids = _CAPI_DGLOmegaToBlock(empty_graph._graph,
                                                    F.zerocopy_to_dgl_ndarray(u),
                                                    F.zerocopy_to_dgl_ndarray(v),
                                                    F.zerocopy_to_dgl_ndarray(dst_ids),
                                                    F.zerocopy_to_dgl_ndarray(src_ids))
    block = DGLBlock(graph_idx, (["_N"], ["_N"]), ["_E"])
    block.srcdata[dgl.NID] = F.zerocopy_from_dgl_ndarray(src_orig_ids)
    return block

def partition_request(
    num_machines,
    num_gpus_per_machine,
    nid_partitions,
    target_gnids,
    target_features,
    src_gnids,
    dst_gnids):

    num_targets = target_gnids.shape[0]
    num_tot_gpus = num_machines * num_gpus_per_machine
    num_assigned_targets_per_gpu = get_num_assigned_targets_per_gpu(
        num_machines, num_gpus_per_machine, num_targets)
    target_features_list = target_features.split(num_assigned_targets_per_gpu)

    num_assigned_targets_per_gpu = torch.tensor(num_assigned_targets_per_gpu, dtype=torch.int64)

    ret = _CAPI_DGLOmegaPartitionRequest(
        num_machines,
        num_gpus_per_machine,
        F.zerocopy_to_dgl_ndarray(nid_partitions),
        F.zerocopy_to_dgl_ndarray(num_assigned_targets_per_gpu),
        F.zerocopy_to_dgl_ndarray(target_gnids),
        F.zerocopy_to_dgl_ndarray(src_gnids),
        F.zerocopy_to_dgl_ndarray(dst_gnids))

    src_gnids_list = []
    dst_gnids_list = []

    for i in range(num_tot_gpus):
        src_gnids_list.append(F.from_dgl_nd(ret[i]))
        dst_gnids_list.append(F.from_dgl_nd(ret[num_tot_gpus + i]))

    return (
        [target_gnids] * num_tot_gpus,
        target_features_list,
        src_gnids_list,
        dst_gnids_list
    )

def sample_edges(target_gnids, src_gnids, dst_gnids, fanouts):
    ret = _CAPI_DGLOmegaSampleEdges(
        F.zerocopy_to_dgl_ndarray(target_gnids),
        F.zerocopy_to_dgl_ndarray(src_gnids),
        F.zerocopy_to_dgl_ndarray(dst_gnids),
        F.zerocopy_to_dgl_ndarray(torch.tensor(fanouts, dtype=torch.int64)))

    return [
        (F.from_dgl_nd(ret[2 * i]), F.from_dgl_nd(ret[2 * i + 1])) for i in range(len(fanouts))
    ]

def trace_gen_helper(
    first_new_gnid,
    infer_target_mask,
    batch_local_ids,
    u,
    v,
    u_in_partitions,
    v_in_partitions,
    independent):
    ret = _CAPI_DGLOmegaTraceGenHelper(
        first_new_gnid,
        F.zerocopy_to_dgl_ndarray(infer_target_mask),
        F.zerocopy_to_dgl_ndarray(batch_local_ids),
        F.zerocopy_to_dgl_ndarray(u),
        F.zerocopy_to_dgl_ndarray(v),
        F.zerocopy_to_dgl_ndarray(u_in_partitions),
        F.zerocopy_to_dgl_ndarray(v_in_partitions),
        independent)

    target_gnids = F.from_dgl_nd(ret[0])
    src_gnids = F.from_dgl_nd(ret[1])
    dst_gnids = F.from_dgl_nd(ret[2])

    return target_gnids, src_gnids, dst_gnids

def collect_cpp_traces():
    return _CAPI_DGLOmegaGetCppTraces()

def partition_facebook_dataset(
    num_parts,
    input_dir,
    edge_file_paths,
    include_out_edges,
    infer_prob,
    num_omp_threads
):
    _CAPI_DGLOmegaPartitionFacebookDataset(
        num_parts,
        input_dir,
        edge_file_paths,
        include_out_edges,
        infer_prob,
        num_omp_threads
    )


_init_api("dgl.omega.omega_apis")
